lib/hterm.js

```diff
@@ -12,6 +13,40 @@ hterm.Terminal.prototype.onMouse_ = function (e) {
   return oldMouse.call(this, e);
 };
 
+function containsNonLatinCodepoints (s) {
+  return /[^\u0000-\u00ff]/.test(s);
+}
+hterm.Terminal.IO.prototype.writeUTF8 = function (string) {
+  if (this.terminal_.io !== this) {
+    throw new Error('Attempt to print from inactive IO object.');
+  }
+  if (containsNonLatinCodepoints(string)) {
+    let length = unicodeStringUtils.length(string);
+    for (var char = 0; char <= length; char++) {
+      this.terminal_.interpret(unicodeStringUtils.slice(string, char, char + 1));
+    }
+  } else {
+    this.terminal_.interpret(string);
+  }
+};
+
+const oldCreateContainer = hterm.TextAttributes.prototype.createContainer;
+hterm.TextAttributes.prototype.createContainer = function (text) {
+  const container = oldCreateContainer.call(this, text);
+  // @TODO: Measure the actual width
+  if (containsNonLatinCodepoints(text) && unicodeStringUtils.length(text) === 1) {
+    container.style.width = '7.23438px';
```

Comment:
codePoint should be relatively easy as it should either be single width or
double width with a monospaced font.
On Tue, Aug 2, 2016 at 10:16 PM James Hall <james@parall.ax> wrote:

> I think I'm going to measure the font on boot to get the width. A cache
> couldn't hurt either for sure.
> On Tue, Aug 2, 2016 at 8:12 PM Guillermo Rauch <notifications@github.com>
> wrote:
>
>> In lib/hterm.js
>> <https://github.com/zeit/hyperterm/pull/535#discussion_r73218490>:
>>
>> > +  if (containsNonLatinCodepoints(string)) {
>> > +    let length = unicodeStringUtils.length(string);
>> > +    for (var char = 0; char <= length; char++) {
>> > +      this.terminal_.interpret(unicodeStringUtils.slice(string, char, char + 1));
>> > +    }
>> > +  } else {
>> > +    this.terminal_.interpret(string);
>> > +  }
>> > +};
>> > +
>> > +const oldCreateContainer = hterm.TextAttributes.prototype.createContainer;
>> > +hterm.TextAttributes.prototype.createContainer = function (text) {
>> > +  const container = oldCreateContainer.call(this, text);
>> > +  // @TODO: Measure the actual width
>> > +  if (containsNonLatinCodepoints(text) && unicodeStringUtils.length(text) === 1) {
>> > +    container.style.width = '7.23438px';
>>
>> we're going to need to cache by fontFamily / fontSize / codepoint correct?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/zeit/hyperterm/pull/535/files/f4b01a89b33ee24f83050009e242319d5b81db30#r73218490>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAC5s9aznUbqBs3uWIQHRpSX9_ptkYMmks5qb5Z7gaJpZM4JZJox>
>> .
>>
> --
> James Hall
> Director
> Parallax Agency Ltd
> M. 07894950320
>
-- 
James Hall
Director
Parallax Agency Ltd
M. 07894950320